File: LC/python/Balanced Binary Tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def calculate_height(self,root):
        if root == None:
            return 0
        else:
            return max(self.calculate_height(root.right),self.calculate_height(root.left))+1
            
    def isBalanced(self, root):
        if root == None:
            return True
        elif abs(self.calculate_height(root.right) - self.calculate_height(root.left)) <= 1:
            return self.isBalanced(root.right) and self.isBalanced(root.left) 
        else:
            return False
        
            
        

File: LC/java/Balanced Binary Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
       
        int h = getHeight(root);
        if(h==-1){
            return false;
        }
        else{
            return true;
        }
    }
    
    private int getHeight(TreeNode x){
        if(x==null){
            return 0;
        }
        
        int l = getHeight(x.left);
        if(l==-1){
            return -1;
        }
        
        int r = getHeight(x.right);
        if(r==-1){
            return -1;
        }
        
        if(Math.abs(r-l)>1){
            return -1;
        }
        else{
            return 1+Math.max(l,r);
        }
    }
}